paddle/fluid/framework/channel.h

```diff
@@ -51,6 +53,77 @@ void CloseChannel(Channel<T>* ch) {
   ch->Close();
 }
 
+/*
+ * The ChannelHolder class serves two main purposes:
+ * 1. It acts as a unified wrapper for the different kinds of
+ *    channels, i.e. Buffered and Unbuffered channels. This is
+ *    similar to the ReaderHolder class.
+ * 2. It also helps us in TypeHiding. This is similar to the
+ *    PlaceHolder implementations in variable.h and tensor.h.
+ */
+class ChannelHolder {
+ public:
+  template <typename T>
+  void Reset(size_t buffer_size) {
+    holder_.reset(new PlaceholderImpl<T>(buffer_size));
+  }
+
+  template <typename T>
+  bool Send(T* data) {
+    if (!IsInitialized()) return false;
+    PADDLE_ENFORCE_EQ(holder_->Type(), std::type_index(typeid(T)));
+    // Static cast should be safe because we have ensured that types are same
+    Channel<T>* channel = static_cast<Channel<T>*>(holder_->Ptr());
+    return channel != nullptr ? channel->Send(data) : false;
+  }
+
+  template <typename T>
+  bool Receive(T* data) {
+    if (!IsInitialized()) return false;
+    PADDLE_ENFORCE_EQ(holder_->Type(), std::type_index(typeid(T)));
+    Channel<T>* channel = static_cast<Channel<T>*>(holder_->Ptr());
+    return channel != nullptr ? channel->Receive(data) : false;
+  }
+
+  void close() {
+    if (IsInitialized()) holder_->Close();
+  }
+
+  inline bool IsInitialized() const { return holder_ != nullptr; }
+
+ private:
+  /**
+   * @note    Placeholder hides type T, so it doesn't appear as a template
+   *          parameter of ChannelHolder.
+   */
+  struct Placeholder {
+    virtual ~Placeholder() {}
+    virtual const std::type_index Type() const = 0;
+    virtual void* Ptr() const = 0;
+    virtual void Close() const = 0;
+    std::type_info type_;
+  };
+
+  template <typename T>
+  struct PlaceholderImpl : public Placeholder {
+    PlaceholderImpl(size_t buffer_size) : type_(std::type_index(typeid(T))) {
+      channel_.reset(MakeChannel<T>(buffer_size));
+    }
+
+    virtual const std::type_index Type() const { return type_; }
+    virtual void* Ptr() const { return static_cast<void*>(channel_.get()); }
+    virtual void Close() {
+      if (channel_) channel_->Close();
+    }
+
+    std::unique_ptr<Channel<T>*> channel_;
```

Comment:
It looks to me that it might be possible to remove the current base class `Channel` from `channel.h`, and rename `ChannelHolder` into `Channel`, if we write here

```c++
std::unique_ptr<
    boost::variant<
        BufferedChannel<T>, 
        UnbufferedChannel<T>>> channel_;
```

I am not an expert of `boost::variant`, but I see it is used many places in our codebase. For example:

https://github.com/PaddlePaddle/Paddle/blob/24509f4af942bb250564756ad636691c7921e1df/paddle/fluid/platform/place.h#L53

Reply:
I tried to verify that we could use the polymorphism provided by boost::variant to call Send:

```c++
#include <iostream>
#include <boost/variant.hpp>

class BufferedChannel {
public:
  bool Send() {
    std::cout << "BufferedChannel::Send\n";
    return false;
  }
};

class UnbufferedChannel {
public:
  bool Send() {
    std::cout << "UnbufferedChannel::Send\n";
    return false;
  }
};

class SendVisitor : public boost::static_visitor<bool> {
public:
  template <typename T>
  bool operator()(T& t) const {
    return t.Send();
  }
};

int main() {
  boost::variant<BufferedChannel, UnbufferedChannel> ch;
  BufferedChannel bch;
  ch = bch;
  boost::apply_visitor(SendVisitor(), ch);
  return 0;
}
```

Reply:
Thank you @wangkuiyi . This looks interesting. I'll look into this Visitor pattern polymorphism.

Reply:
By extending the above example to work with unique_ptr:

```c++
#include <iostream>
#include <boost/variant.hpp>

class BufferedChannel {
public:
  bool Send() {
    std::cout << "BufferedChannel::Send\n";
    return false;
  }
};

class UnbufferedChannel {
public:
  bool Send() {
    std::cout << "UnbufferedChannel::Send\n";
    return false;
  }
};

class SendVisitor : public boost::static_visitor<bool> {
public:
  template <typename T>
  bool operator()(T* t) const {
    return t->Send();
  }
};

int main() {
  typedef boost::variant<BufferedChannel*, UnbufferedChannel*> ChannelVariant;
  std::unique_ptr<ChannelVariant> ch;
  ch.reset(new ChannelVariant(new BufferedChannel));
  boost::apply_visitor(SendVisitor(), *ch.get());
  return 0;
}
```